### ISSUE TITLE
Repository config file permissions wrong when using ACLs

### ISSUE THREAD

**Issue by PromyLOPh** [borgbackup/borg#6933]
## Have you checked borgbackup docs, FAQ, and open Github issues?

Yes

## Is this a BUG / ISSUE report or a QUESTION?

BUG

#### Your borg version (borg -V).

.borg-real 1.2.1

#### Operating system (distribution) and version.

GNU Guix (Linux), commit d5ba31c19accdf14bca23ebc0c3baf1c5b08fb5f

#### Hardware / network configuration, and filesystems used.

Any filesystem with ACL support (ext4, tmpfs, …)

## Describe the problem you're observing.

```console
$ umask
022
$ mkdir withAcl
$ setfacl -m d:g:audio:rx withAcl
$ borg init -e none withAcl/repo
$ ls -lh withAcl/repo/config
-rw-------+ 1 user user 209  4. Aug 11:36 withAcl/repo/config
$ getfacl withAcl/repo/config
# file: withAcl/repo/config
# owner: user
# group: user
user::rw-
group::r-x			#effective:---
group:audio:r-x			#effective:---
mask::---
other::---
```

borg uses `mkstemp` to create its repository config file, which apparently clears the mask for ACLs. Thus the inherited group `audio` (and even the standard file group) does not have any permissions on that file.

The following patch fixes the issue, but I’m not sure it applies to Windows/Darwin/BSD/… as well:

```patch
diff -Naur borgbackup-1.2.1.orig/src/borg/platform/base.py borgbackup-1.2.1/src/borg/platform/base.py
--- borgbackup-1.2.1.orig/src/borg/platform/base.py	1970-01-01 01:00:01.000000000 +0100
+++ borgbackup-1.2.1/src/borg/platform/base.py	2022-08-04 11:18:06.205481514 +0200
@@ -227,7 +227,14 @@

     def __enter__(self):
         from .. import platform
-        self.tmp_fd, self.tmp_fname = tempfile.mkstemp(prefix=self.tmp_prefix, suffix='.tmp', dir=self.dir)
+        i = 0
+        while True:
+            self.tmp_fname = os.path.join(self.dir, self.tmp_prefix + str(i) + '.tmp')
+            try:
+                self.tmp_fd = os.open(self.tmp_fname, flags=os.O_RDWR|os.O_CREAT|os.O_EXCL, mode=0o666)
+                break
+            except FileNotFoundError:
+                i += 1
         self.f = platform.SyncFile(self.tmp_fname, fd=self.tmp_fd, binary=self.binary)
         return self.f

@@ -239,14 +246,6 @@
             safe_unlink(self.tmp_fname)  # with-body has failed, clean up tmp file
             return  # continue processing the exception normally

-        # tempfile.mkstemp always uses owner-only file permissions for the temp file,
-        # but as we'll rename it to the non-temp permanent file now, we need to respect
-        # the umask and change the file mode to what a normally created file would have.
-        # thanks to the crappy os.umask api, we can't query the umask without setting it. :-(
-        umask = os.umask(UMASK_DEFAULT)
-        os.umask(umask)
-        os.chmod(self.tmp_fname, mode=0o666 & ~ umask)
-
         try:
             os.replace(self.tmp_fname, self.path)  # POSIX: atomic rename
         except OSError:
```

```console
$ ls -lh withAcl/repo2/config
-rw-r--r--+ 1 user user 209  4. Aug 11:41 withAcl/repo2/config
$ getfacl withAcl/repo2/config
# file: withAcl/repo2/config
# owner: user
# group: user
user::rw-
group::r-x			#effective:r--
group:audio:r-x			#effective:r--
mask::r--
other::r--
```

**Comment by ThomasWaldmann** [comment]
So, guess we're now at reimplementing python's `tempfile.mkstemp` in a (for our usage) less problematic way.

https://github.com/python/cpython/blob/d4792ce916b94d090b6c7bce8b0f973e840c9e4e/Lib/tempfile.py#L305

Some of the stuff it does might be relevant for security, so guess this needs:

- a new function, e.g. in `helpers.fs.borg_mkstemp`
- a review of its security properties

**Comment by PromyLOPh** [comment]
Oh, I didn’t know it was implemented in pure Python – thought it was just a call to the libc’s mkstemp. The only real difference seems to be Python’s hard-coded `0o600` permissions on `open` then. Bummer that Python does not allow overriding it.

**Comment by ThomasWaldmann** [comment]
Filed an issue on the cpython issue tracker. ^^^

**Comment by PromyLOPh** [comment]
Yes, the fix in #6936 solves this ACL issue for me. Thank you very much!
